packages/qri-node/lib/qri-node.js

```diff
@@ -1,25 +1,41 @@
 const fs = require('fs-extra')
-const { spawn } = require('child_process');
+const { spawnSync } = require('child_process');
 
 module.exports = {
+  /**
+  * 'initialize' creates a new dataset, links it to the current directory,
+  * and creates starter files for the dataset's components
+  * @param {string} dirPath - The path to create the new dataset in
+  */
   init: (dirPath) => {
     console.log(`initializing Qri Dataset in ${dirPath}`)
     // make an empty dir
     fs.emptyDirSync(dirPath)
 
     return new Promise((resolve, reject) => {
       try {
-        const child = spawn(`cd ${dirPath} && qri init`, [], { shell: true, stdio: [process.stdin, process.stdout, process.stderr] })
-        child.on('exit', () => {
-          resolve()
-        })
+        const { status, stdout, stderr } = spawnSync(`cd ${dirPath} && qri init`, [], { shell: true })
+
+        if (status === 1) {
+          reject(new Error(stderr.toString().trim()))
+        } else {
+          resolve(stdout.toString().trim())
+        }
 
       } catch(e) {
         reject(e)
       }
     })
   },
 
+  /**
+  * saves a dataset
+  * @param {string} dsRef - The dataset reference to save
+  * @param {Object} saveOptions
+  * @param {string} saveOptions.body Path to the body file to be saved
+  * @param {Array.<string>} saveOptions.file Array of paths to additional files (meta, readme, etc) to be saved
+  * @returns {Promise} - If resolved, stdout, if rejected, stderr
+  */
   save: (dsRef, { body, file }) => {
```

Comment:
None of these functions are doing validation on `dsRef`, so it's possible for a user to pass a string to any of them that ends up injecting extra commands into the shell. `qri.save("me/my_dataset; cat /etc/passwd")` for example.

Luckily, dsRef has a well-defined grammar (https://github.com/qri-io/qri/blob/master/dsref/parse.go), so it's not too hard to add some simple validation, and throw an error if the validation fails. For example, the python library uses the regex seen here: https://github.com/qri-io/qri-python/blob/master/qri/dsref.py

Reply:
will add validation